py/desispec/sky.py

```diff
@@ -655,6 +666,104 @@ def fit_and_interpolate(delta,skyfibers,mean,components,label="") :
     cskyflux[:,bad]=0.
     modified_cskyivar[:,bad]=0.
 
+    if fit_offsets :
+
+        cfinder = CalibFinder([frame.meta])
+        amps    = get_amp_ids(frame.meta)
+
+        psf_filename = findfile('psf',frame.meta["NIGHT"],frame.meta["EXPID"],frame.meta["CAMERA"])
+        if not os.path.isfile(psf_filename) :
+            log.error("No PSF file "+psf_filename)
+            raise IOError("No PSF file "+psf_filename)
+        log.info("Using PSF {}".format(psf_filename))
+        tset = read_xytraceset(psf_filename)
+
+        tmp_fibers = np.arange(frame.nspec)
+        tmp_x = np.zeros(frame.flux.shape,dtype=float)
+        tmp_y = np.zeros(frame.flux.shape,dtype=float)
+        for fiber in tmp_fibers :
+            tmp_x[fiber] = tset.x_vs_wave(fiber=fiber,wavelength=frame.wave)
+            tmp_y[fiber] = tset.y_vs_wave(fiber=fiber,wavelength=frame.wave)
+
+        sectors=[]
+        for amp in amps :
+            key="OFFCOLS"+amp
+            sec = parse_sec_keyword(frame.meta['CCDSEC'+amp])
+            yb=sec[0].start
+            ye=sec[0].stop
+
+            if cfinder.haskey(key) :
+                val=cfinder.value(key)
+                for tmp1 in val.split(",") :
+                    tmp2=tmp1.split(":")
+                    if len(tmp2)!=2 :
+                        mess="cannot decode {}={}".format(key,val)
+                        log.error(mess)
+                        raise KeyError(mess)
+                    xb=max(sec[1].start,int(tmp2[0]))
+                    xe=min(sec[1].stop,int(tmp2[1]))
+                    sector = [yb,ye,xb,xe]
+                    sectors.append( sector )
+                    log.info("Adding CCD sector in amp {} with offset: {}".format(amp,sector))
+
+        if len(sectors)>0 :
+            # fit as many parameters as twice the number of sectors
+            # (one offset for fibers/wavelength in the sector and one for the others to compensate)
+
```

Comment:
Despite the clear commenting, I don't get this.  I expect an offset in each sector and an offset outside all the sectors, so npar = nsector + 1.  Maybe len(sectors) is always zero or 1, where this difference doesn't matter.  But the code is clearly designed to handle larger numbers of sectors.  I guess it just makes ill-conditioned matrices if I've understood it correctly and len(sectors) > 1, but I probably just haven't understood it.

Reply:
I considered an offset in each sector and an offset for the outside of each sector as well. And sectors and their outside can overlap.

Reply:
I'm worried that A won't have an inverse whenever there are two or more sectors?  I hadn't thought about overlapping sectors, but add long as the sectors are rectangular I think you'll have a nontrivial null space whenever there is more than one sector.

Reply:
> I think you are right. I also need to revise the code for 2 other reasons: the 'outside' of a sector should contain only the pixels addressing the same wavelength range (this is a real shortcoming in the current version). And more generally, because the wavelength are not aligned with the pixel grid, I should be doing the fit differently (either 2 pass sky subtraction which is less elegant but less invasive for the current implementation, or incorporate the fit of offsets in a simultaneous fit with the sky spectrum).

Reply:
Checking in: @julienguy @schlafly has this concern been resolved?

Reply:
by slack @julienguy says this is resolved